torch/utils/data/dataloader.py

```diff
@@ -55,131 +81,84 @@ class DataLoader(object):
-                raise ValueError('batch_sampler option is mutually exclusive '
-                                 'with batch_size, shuffle, sampler, and '
-                                 'drop_last')
-            self.batch_size = None
-            self.drop_last = None
-
-        if sampler is not None and shuffle:
-            raise ValueError('sampler option is mutually exclusive with '
-                             'shuffle')
-
-        if self.num_workers < 0:
-            raise ValueError('num_workers option cannot be negative; '
-                             'use num_workers=0 to disable multiprocessing.')
-
-        if batch_sampler is None:
+        self.convert_fn = convert_fn
+        self.collate_fn = collate_fn
+        if isinstance(dataset, IterableDataset):
+            # FIXME: check default for some args
+            self.mode = _DataLoaderStrategy.Iterable
+            self.sampler = None
+        elif batch_size is None:
+            # FIXME: check default for some args
+            self.mode = _DataLoaderStrategy.Map
             if sampler is None:
                 if shuffle:
                     sampler = RandomSampler(dataset)
                 else:
                     sampler = SequentialSampler(dataset)
-            batch_sampler = BatchSampler(sampler, batch_size, drop_last)
-
-        self.sampler = sampler
-        self.batch_sampler = batch_sampler
+            self.sampler = sampler
+        else:
+            # FIXME: check default for some args
+            self.mode = _DataLoaderStrategy.MapWithBatchedRead
+
+            self.batch_size = batch_size
+            self.drop_last = drop_last
+
+            if timeout < 0:
+                raise ValueError('timeout option should be non-negative')
+
+            if batch_sampler is not None:
+                if batch_size > 1 or shuffle or sampler is not None or drop_last:
+                    raise ValueError('batch_sampler option is mutually exclusive '
+                                     'with batch_size, shuffle, sampler, and '
+                                     'drop_last')
+                self.batch_size = None
+                self.drop_last = None
+
+            if sampler is not None and shuffle:
+                raise ValueError('sampler option is mutually exclusive with '
+                                 'shuffle')
+
+            if self.num_workers < 0:
+                raise ValueError('num_workers option cannot be negative; '
+                                 'use num_workers=0 to disable multiprocessing.')
+
+            if batch_sampler is None:
+                if sampler is None:
+                    if shuffle:
+                        sampler = RandomSampler(dataset)
+                    else:
+                        sampler = SequentialSampler(dataset)
+                batch_sampler = BatchSampler(sampler, batch_size, drop_last)
+
+            self.sampler = batch_sampler
```

Comment:
Looks like the `TypeError: unhashable type: 'list'` error stems from this change. Code expects the sampler to return List[int], but it actually returns List[List[int]] since sampler is initialized to BatchSampler whereas earlier it would be initialized to RandomSampler.

Reply:
oh i see. okay, i'll fix the type hints as well!

Reply:
hmm wait. `.sampler` isn't even typed in `dataloader.pyi`. Are some call sites just accessing nonpublic members and enforcing types on them?.... Those internal code should be fixed as well.

Reply:
I agree those should be fixed, although we can't change the behavior to initialize sampler to BatchSampler since the code relies on the behavior of RandomSampler (not the type).

Reply:
Can we fix the default behavior here?

Reply:
I don't know why this needs fixing and I don't quite understand your argument about `RandomSampler` and `BatchSampler`. This does not change the data loader's behavior whatsoever. 

IIUC, this is an issue where some downstream code uses these private attributes and expects them to behave whatever they expect it to be. For such cases, I strongly don't agree that we need to fix. If they hack our library and decide to have test relying on the hack, they should be the one fixing. I have suffered from such things already before when doing the dataloader refactoring.

These said, I'm now not at facebook so I can't yell at them. The new commit should hopefully "unbreak" their broken test.

Reply:
It would be cool if you can tell me if this does unbreak the test.

Reply:
Any updates? @pritamdamania87

Reply:
Sorry I missed this message earlier. Yes, the latest changes do fix the tests. Thanks a lot!